onnxconverter_common/auto_mixed_precision_model_path.py

```diff
@@ -0,0 +1,250 @@
+        atol = 1e-8
+
+    if rtol is None and validate_fn is None:
+        raise ValueError("Argument `validate_fn` and `rtol` cannot both be `None`.")
+
+    if location is None:
+        location = "fp16_tensor.data"
+
+    tmp_model32_path, tmp_model32_tensor_name = generate_temp_filename(target_model_path)
+
+    kwargs = {
+        "tmp_model32_path": tmp_model32_path,
+        "tmp_model32_tensor_name": tmp_model32_tensor_name,
+        "source_model_path": source_model_path,
+        "input_feed": input_feed,
+        "target_model_path": target_model_path,
+        "location": location,
+        "validate_fn": validate_fn,
+        "rtol": rtol,
+        "atol": atol,
+        "keep_io_types": keep_io_types,
+        "providers": providers
+        }
+
+    print("copy source model to temp folder, change to external data, then check.")
+    model_32, output_32 = copy_fp32_model(**kwargs)
+
+    print("convert to initial fp16 model, then check.")
+    node_names = [n.name for n in model_32.graph.node if n.op_type not in ["Loop", "If", "Scan"]]
+    kwargs["model_32"] = model_32
+    kwargs["res1"] = output_32
+    kwargs["node_block_list"] = node_names
+    kwargs["return_model"] = False
+    run_attempt(**kwargs)
+
+    print("Sanity checks passed. Starting autoconvert.")
+    final_nodes = try_to_convert_to_valid_fp16_model(**kwargs)
+
+    print(".final convert.")
+    kwargs["node_block_list"] = final_nodes
+    kwargs["return_model"] = True
+    valid, model = run_attempt(**kwargs)
+    if not valid:
+        raise ValueError("validation failed for final fp16 model")
+    print("Final model validated successfully.")
+
+    clean_output_folder(**kwargs)
+
+
+def try_to_convert_to_valid_fp16_model(**kwargs):
+    node_names = kwargs.get('node_block_list')
+
+    segments = SegmentList(node_names)
+    i = 0
+    while segments.get_largest() is not None:
+        seg = segments.get_largest()
+        nodes_to_try = segments.get_nodes(seg)
+        i += 1
+        print("Running attempt %d excluding conversion of %s nodes" % (i, len(nodes_to_try)))
+        kwargs["node_block_list"] = nodes_to_try
+        if run_attempt(**kwargs):
+            seg.good = True
+            print("Attempt succeeded.")
+        else:
+            print("Attempt failed.")
+            if seg.size == 1:
+                seg.bad = True
+            else:
+                seg.split()
+        print("segments=", segments)
+    print("Done! these nodes will keep float32 type:", segments.get_nodes())
+
+    return segments.get_nodes()
+
+
+def generate_temp_filename(target_model_path):
+    target_model_folder = os.path.dirname(target_model_path)
+    if not os.path.exists(target_model_folder):
+        os.mkdir(target_model_folder)
+    tensor_filename = str(uuid.uuid1())
+    onnx_filename = os.path.join(target_model_folder, tensor_filename + ".onnx")
+    return onnx_filename, tensor_filename + ".data"
+
+
+def copy_fp32_model(**kwargs):
```

Comment:
Agree that kwargs will give us more flexibility, but it also confuses users, and other authors in future, with the parameters of this function.

I suggest changing the signature of this function to normal parameters. Or, at least, change it to an internal function, like "_copy_fp32_model", so that the users won't call it directly.

Reply:
resolved